### PR TITLE
First commit for review

### PR DESCRIPTION
**Pull request by Radicals27**

### CODE REVIEW

**Comment by Radicals27** [comment]
Just checking if this is still OK?

File: CodingTracker.Radicals27/DBController.cs [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+
+        internal static void InitialiseDB()
+        {
+            using (var connection = new SqliteConnection(completeConnectionString))
+            {
+                connection.Open();
+                var tableCmd = connection.CreateCommand();
+
+                tableCmd.CommandText =
+                    @"CREATE TABLE IF NOT EXISTS hours_playefod (
```
Comment (chrisjamiecarter) [inline CodingTracker.Radicals27/DBController.cs]: 🟡 **Typo**

❓ Should this be `hours_played`?

File: CodingTracker.Radicals27/DBController.cs [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+                return SessionStartTime.ToString();
+            }
+            else
+            {
+                Console.WriteLine("No session has begun.");
+                return "";
+            }
+        }
+
+        internal static void InitialiseDB()
```
Comment (chrisjamiecarter) [inline CodingTracker.Radicals27/DBController.cs]: 🟠 **Dapper?**

💡 This project requires you to use Dapper, I don't know if this was a previous method copied over, or you just forgot for this method, but it should have been implemented like the below.

```
using var connection = new SqliteConnection(completeConnectionString);
var createSql = @"CREATE TABLE IF NOT EXISTS hours_played (
            Id INTEGER PRIMARY KEY AUTOINCREMENT,
            Date DATE,
            StartTime INTEGER,
            EndTime INTEGER
            )";

connection.Execute(createSql);

var checkSql = @"SELECT COUNT(*) FROM hours_played";
var count = connection.QuerySingle<int>(checkSql);
if (count == 0)
{
    SeedDatabase(connection);
}
```

File: CodingTracker.Radicals27/DBController.cs [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+                DateTime startTime = new DateTime(1, 1, 1, randomStartHour, randomStartMinute, 0);
+                int randomDurationMinutes = random.Next(minDuration, maxDuration);
+
+                DateTime endTime = startTime.AddMinutes(randomDurationMinutes);
+
+                int formattedStartTime = int.Parse(startTime.ToString("HHmm"));
+                int formattedEndTime = int.Parse(endTime.ToString("HHmm"));
+
+                // Dapper insertion
+                string sql = "INSERT INTO hours_played (Date, StartTime, EndTime) VALUES (@Date, @StartTime, @EndTime)";
```
Comment (chrisjamiecarter) [inline CodingTracker.Radicals27/DBController.cs:92]: 🟢 **SQL Parameters**

⭐

File: CodingTracker.Radicals27/DBController.cs [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+
+                List<CodingSession> allRecords = connection.Query<CodingSession>(sql).ToList();
+
+                connection.Close();
+
+                return allRecords;
+            }
+        }
+
+        internal static void Insert()
```
Comment (chrisjamiecarter) [inline CodingTracker.Radicals27/DBController.cs:120]: 🟠 **Tight Coupling**

💡 You have tightly coupled your database logic with your UI logic. To uncouple, you could do the UI get inputs before calling this method, then call this method and add the values you need as parameters.

i.e.
```
internal static void Insert(string? date, int?  startTime, int? endTime) { ... }
```

File: CodingTracker.Radicals27/DBController.cs [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+                {
+                    Date = date,
+                    StartTime = startTime,
+                    EndTime = endTime,
+                    RecordId = recordId
+                });
+
+                Console.Clear();
+
+                if (affectedRows > 0)
```
Comment (chrisjamiecarter) [inline CodingTracker.Radicals27/DBController.cs:227]: 🟠 **DB & UI**

💡 Again to uncouple, you could have made the return type of this method a `bool` or `int`, and handled the `Console.WriteLine` in the UI.

```
internal static bool Update()
{
    ...

    return affectedRows > 0;
}
```

or


```
internal static int Update()
{
    ...

    return affectedRows;
}
```

File: CodingTracker.Radicals27/DBController.cs [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+                codingSessionCount = Convert.ToInt32(reportCmd.ExecuteScalar());
+
+                connection.Close();
+            }
+
+            Console.Clear();
+            Console.WriteLine($"The habit was performed {codingSessionCount} times in {yearInput}.");
+        }
+
+        private static string GenerateRandomDate(Random random)
```
Comment (chrisjamiecarter) [inline CodingTracker.Radicals27/DBController.cs:276]: 🟡 **Random.Shared**

💡 You can use `Random.Shared` to generate a random number without needing to instantiate an object.

File: CodingTracker.Radicals27/Readme.md
Comment (chrisjamiecarter) [inline CodingTracker.Radicals27/Readme.md:1]: 🟢 **Readme File**

⭐ Excellent readme file

File: CodingTracker.Radicals27/UserInput.cs [hunk 7]
```diff
@@ -0,0 +1,142 @@
+using System.Globalization;
+
+namespace coding_tracker
+{
+    /// <summary>
+    /// Responsible for taking in all user input and validating it
+    /// </summary>
+    class UserInput
```
Comment (chrisjamiecarter) [inline CodingTracker.Radicals27/UserInput.cs:8]: 🟠 **Method Accessibility**

💡 Always try to explicitly declare accessibility. public, internal, private, etc

**Review by chrisjamiecarter — CHANGES_REQUESTED** [review]
Hey @Radicals27 👋 ,

Excellent work on your CodingTracker project submission 🎉!

I have performed a peer review and have added some comments. Review/ignore as you wish.

***

However, your project has the following issues 🔴 !
Please revisit to try understand and fix these issues 🔧 , as they relate to the projects requirements and they block my approval.

🔴 **Insert manual record**
Steps to reproduce:
- Click `2. Insert a record`
- Add data `01-01-24`, `0900`, `1000`
- Click `1. View all records`

image


***

Feel free to reach out if you have any questions or if you'd like to collaborate further on any of these points 🆘 .

I will go ahead and wait for you to make your changes, commit and push (which will then notify me) 😊 !

Thanks,
@chrisjamiecarter 👍

File: CodingTracker.Radicals27/DBController.cs (hunk 1 above)
Comment (Radicals27) [reply CodingTracker.Radicals27/DBController.cs]: Fixed in next commit.

As for the above issue - adding a new entry "01-01-24" , "0900", "1000"
I cannot replicate the issue, i.e it displays all records when viewing all of them after this insert.
I notice in your screenshot that the line throwing the error is line 131, however that is line 114 in my DBController.cs code so maybe there was code deleted on your end?

File: CodingTracker.Radicals27/DBController.cs (hunk 2 above)
Comment (Radicals27) [reply CodingTracker.Radicals27/DBController.cs]: Fixed in next commit.

**Comment by Radicals27** [comment]
OK I wasn't able to replicate the exception in your first message but fixed a couple of the issues, please give me more info on the exception, as I can't replicate it.  Ta

**Review by chrisjamiecarter — CHANGES_REQUESTED** [review]
Hey @Radicals27 👋 ,

Great job on getting your updates pushed 💪 ,

I am still getting the error.

I have done some debugging for you and it might be cultural (for reference I am in UK).

Also when inspecting your database file, I notice that the data you seed, and the data you insert is different.

image

Could you give it another go? Or at least prove to me that you can insert a manual record with today's date?

***

Feel free to reach out if you have any questions or if you'd like to collaborate further on any of these points 🆘 .

I will go ahead and wait for you to make your changes, commit and push (which will then notify me) 😊 !

Thanks,
@chrisjamiecarter 👍

**Comment by Radicals27** [comment]
OK I think I have fixed it , there was a discrepency between the en-US date times, one had it and one did not.


Please note I also added a line that firstly deletes the table, to ensure a clean start (on each app load).

Here is some screengrabs showing that I don't get the bug.

**Review by chrisjamiecarter — APPROVED** [review]
😆 Ok, well it is still not working my end. But I will have to trust that it works your end.


I will go ahead and mark as approved, keep up the excellent work on the next projects! 😊

Best regards,
@chrisjamiecarter 👍

**Pull request closed without merging**